lib/content_consistency_checker.rb

```diff
@@ -3,92 +3,124 @@
+      edition = Edition.find_by(
+        content_store: content_store,
+        base_path: content_item.base_path
+      )
+      next if edition
 
-    errors
+      errors[content_item.base_path] << "No edition available."
+    end
   end
 
 private
 
-  attr_reader :content_id, :locale, :ignore_recent
+  attr_reader :content_store, :content_dump, :checked_content
+  attr_writer :errors
 
-  def check_edition(prefix, edition, content_store)
-    return unless edition.base_path
+  def unchecked_content
+    Set.new(content_dump.keys) - checked_content
+  end
+
+  def editions_to_check
+    Edition
+      .with_document
+      .where(content_store: content_store)
+  end
+
+  def get_content_item(path)
+    begin
+      path = path.to_sym
+      content_item = content_dump.fetch(path)
+      checked_content.add?(path)
+      content_item
+    rescue KeyError
+      nil
+    end
+  end
+
+  def hash_field(object)
+    Digest::SHA1.hexdigest(
+      JSON.generate(object)
+    )
+  end
+
+  def draft?
+    content_store == "draft"
+  end
 
-    return if ignore_recent && edition.updated_at < 1.day.ago
+  def check_edition(edition)
+    return unless edition.base_path
 
     path = edition.base_path
 
-    begin
-      content_item = content_store.content_item(path).parsed_content
-    rescue GdsApi::HTTPForbidden
-      # nothing
-    rescue GdsApi::ContentStore::ItemNotFound
-      errors << "#{prefix} #{path} content is missing from the content store."
-    rescue GdsApi::HTTPGone
-      if edition.gone?
-        return
-      else
-        errors << "#{prefix} #{path} is gone in the content store."
+    content_item = get_content_item(edition.base_path)
+
+    unless content_item
+      unless edition.gone?
+        errors[path] << "Content is missing from the content store."
       end
+
+      return
     end
 
-    return if content_item.nil?
+    return if edition.updated_at > content_item.updated_at
 
     if edition.redirect?
-      if content_item["document_type"] != "redirect" ||
-          content_item["schema_name"] != "redirect"
-        errors << "#{prefix} content is not a redirect in the content store."
+      unless content_item.redirect?
+        errors[path] << "Content is not a redirect in the content store."
       end
     elsif edition.gone?
-      if content_item["document_type"] != "gone" ||
-          content_item["schema_name"] != "gone"
-        errors << "#{prefix} content is not gone in the content store."
+      unless content_item.gone?
+        errors[path] << "Content is not gone in the content store."
       end
     else
-      fields = [:rendering_app, :publishing_app, :schema_name, :document_type]
+      fields = [:content_id, :locale, :rendering_app, :publishing_app, :schema_name, :document_type]
       fields.each do |field|
         edition_value = edition.send(field)
-        content_item_value = content_item[field.to_s]
+        content_item_value = content_item.send(field)
```

Comment:
I think `public_send` would be better as we only want to use public methods, and when you see `send` in ruby it tends to trigger a concern that someones trying to use a protected/private method